lib/call-cabal-project-to-nix.nix

```diff
@@ -51,13 +58,93 @@ in
     then throw "provided sha256 for index-state ${index-state-found} is null!" else true);
 
 let
+  span = pred: list:
+    let n = pkgs.lib.lists.foldr (x: acc: if pred x then acc + 1 else 0) 0 list;
+    in { fst = pkgs.lib.lists.take n list; snd = pkgs.lib.lists.drop n list; };
+
+  # Parse lines of a source-repository-package block
+  parseBlockLines = blockLines: builtins.listToAttrs (builtins.concatMap (s:
+    let pair = builtins.match " *([^:]*): *(.*)" s;
+    in pkgs.lib.optional (pair != null) (pkgs.lib.attrsets.nameValuePair
+          (builtins.head pair)
+          (builtins.elemAt pair 1))) blockLines);
+
+  fetchRepo = repo: (pkgs.fetchgit {
+    url = repo.location;
+    rev = repo.tag;
+    sha256 = repo."--sha256";
+  }) + (if repo.subdir or "" == "" then "" else "/" + repo.subdir);
+
+  # Parse a source-repository-package and fetch it if it containts
+  # a line of the form
+  #   --shar256: <<SHA256>>
+  parseBlock = block:
+    let
+      x = span (pkgs.lib.strings.hasPrefix " ") (pkgs.lib.splitString "\n" block);
+      attrs = parseBlockLines x.fst;  
+    in
+      if attrs."--sha256" or "" == ""
+        then {
+          sourceRepo = [];
+          otherText = "\nsource-repository-package\n" + block;
+        }
+        else {
+          sourceRepo = [ (fetchRepo attrs) ];
+          otherText = pkgs.lib.strings.concatStringsSep "\n" x.snd;
+        };
+
+  # Deal with source-repository-packages in a way that will work on
+  # hydra build agents (as long as a sha256 is included).
```

Comment:
I think this should be "will work in restricted-eval mode". As that's essentially what's happening.